bee-network/bee-autopeering/src/discovery/query.rs

```diff
@@ -0,0 +1,267 @@
+
+                    manager::remove_peer_from_active_list(
+                        &peer_id,
+                        &ctx_.master_peers,
+                        &ctx_.active_peers,
+                        &ctx_.replacements,
+                        &ctx_.event_tx,
+                    )
+                }
+            });
+        } else {
+            log::debug!("Currently no peers to reverify.");
+        }
+    })
+}
+
+// Hive.go: returns the oldest peer, or nil if empty.
+fn peer_to_reverify(active_peers: &ActivePeersList) -> Option<PeerId> {
+    active_peers.read().get_oldest().map(|p| *p.peer_id())
+}
+
+// Hive.go:
+// The current strategy is to always select the latest verified peer and one of
+// the peers that returned the most number of peers the last time it was queried.
+pub(crate) fn do_query() -> Repeat<QueryContext> {
+    Box::new(|ctx| {
+        let peers = select_peers_to_query(&ctx.active_peers);
+        if peers.is_empty() {
+            log::warn!("No peers to query.");
+        } else {
+            log::debug!("Querying {} peer/s...", peers.len());
+
+            for peer_id in peers.into_iter() {
+                let ctx_ = ctx.clone();
+
+                // TODO: introduce `UnsupervisedTask` type, that always finishes after a timeout.
+                tokio::spawn(async move {
+                    if let Some(peers) =
+                        manager::begin_discovery(&peer_id, &ctx_.active_peers, &ctx_.request_mngr, &ctx_.server_tx)
+                            .await
+                    {
+                        log::debug!("Query successful. Received {} peers.", peers.len());
+                    } else {
+                        log::debug!("Query unsuccessful. Removing peer {}.", peer_id);
+
+                        manager::remove_peer_from_active_list(
+                            &peer_id,
+                            &ctx_.master_peers,
+                            &ctx_.active_peers,
+                            &ctx_.replacements,
+                            &ctx_.event_tx,
+                        )
+                    }
+                });
+            }
+        }
+    })
+}
+
+// Hive.go: selects the peers that should be queried.
+fn select_peers_to_query(active_peers: &ActivePeersList) -> Vec<PeerId> {
+    let mut verif_peers = manager::get_verified_peers(active_peers);
+
+    // If we have less than 3 verified peers, then we use those for the query.
+    if verif_peers.len() < 3 {
+        verif_peers.into_iter().map(|ap| *ap.peer_id()).collect::<Vec<_>>()
+    } else {
+        // Note: this macro is useful to remove some noise from the pattern matching rules.
+        macro_rules! num {
+            ($t:expr) => {
+                // Panic: we made sure, that unwrap is always okay.
+                $t.as_ref().unwrap().metrics().last_new_peers()
+            };
+        }
+
+        fn len<T>(o: &(Option<T>, Option<T>, Option<T>)) -> usize {
```

Comment:
Isn't this function always going to return 3 in this context ?

Reply:
yup, didn't realize that it became redundant when I refactored this algo. :+1:

Reply:
Okay....it's not actually const 3, but this: `let len = verif_peers.len().min(3);` due to the rule of always including the first peer in the query. Luckily the unit test caught it :sweat_smile: